src/vs/base/browser/performance.ts

```diff
@@ -0,0 +1,231 @@
+	const enum EventPhase {
+		Before = 0,
+		InProgress = 1,
+		Finished = 2
+	}
+	const state = {
+		keydown: EventPhase.Before,
+		input: EventPhase.Before,
+		render: EventPhase.Before,
+		selection: EventPhase.Before
+	};
+
+	/**
+	 * Mark the start of the keydown event.
+	 */
+	export function markKeydownStart() {
+		performance.mark('inputlatency/start');
+		performance.mark('keydown/start');
+		state.keydown = EventPhase.InProgress;
+		queueMicrotask(() => markKeydownEnd());
+	}
+
+	/**
+	 * Mark the end of the keydown event.
+	 */
+	function markKeydownEnd() {
+		// Only measure the first render after keyboard input
+		performance.mark('keydown/end');
+		state.keydown = EventPhase.Finished;
+	}
+
+	/**
+	 * Mark the start of the input event.
+	 */
+	export function markInputStart() {
+		performance.mark('input/start');
+		state.input = EventPhase.InProgress;
+	}
+
+	/**
+	 * Mark the end of the input event.
+	 */
+	export function markInputEnd() {
+		queueMicrotask(() => {
+			performance.mark('input/end');
+			state.input = EventPhase.Finished;
+		});
+	}
+
+	/**
+	 * Mark the start of the animation frame performing the rendering.
+	 */
+	export function markRenderStart() {
+		// Render may be triggered during input, but we only measure the following animation frame
+		if (state.keydown === EventPhase.Finished && state.input === EventPhase.Finished && state.render === EventPhase.Before) {
+			// Only measure the first render after keyboard input
+			performance.mark('render/start');
+			state.render = EventPhase.InProgress;
+			queueMicrotask(() => markRenderEnd());
+		}
+	}
+
+	/**
+	 * Mark the end of the animation frame performing the rendering.
+	 *
+	 * An input latency sample is complete when both the textarea selection change event and the
+	 * animation frame performing the rendering has triggered.
+	 */
+	function markRenderEnd() {
+		performance.mark('render/end');
+		state.render = EventPhase.Finished;
+		record();
+	}
+
+	/**
+	 * Mark when the editor textarea selection change event occurs.
+	 *
+	 * An input latency sample is complete when both the textarea selection change event and the
+	 * animation frame performing the rendering has triggered.
+	 */
+	export function markTextareaSelection() {
+		state.selection = EventPhase.Finished;
+		record();
+	}
+
+	/**
+	 * Record the input latency sample if it's ready.
+	 */
+	function record() {
+		// Skip recording this frame if the buffer is full
+		if (measurementsCount >= Constants.bufferLength) {
+			return;
+		}
+		// Selection and render must have finished to record
+		if (state.selection !== EventPhase.Finished || state.render !== EventPhase.Finished) {
+			return;
+		}
+		// Finish the recording, use a timer to ensure that layout/paint is captured. setImmediate
+		// is used if available (Electron) to get slightly more accurate results
+		('setImmediate' in window ? (window as any).setImmediate : setTimeout)(() => {
```

Comment:
This gives on my machine a compile error.

Reply:
I see CI failed as well, should I just do away with `setImmediate` or do you have a preferred way to workaround the problem here? The downside of using `setTimeout` is other timers could fire after the render which would delay the final recording

Reply:
Why do we need a `setImmediate`/`setTimeout` call here since this code would read the markers which are set deterministically by the other functions?

Reply:
There are other timeouts installed during render which won't get flushed yet, my understanding is setImmediate would fire before those timeouts and setTimeout would fire after

Reply:
> There are other timeouts installed during render

What do you mean? AFAIK the editor doesn't use `setTimeout` during render to render things later. i.e. AFAIK all rendering in the editor is done during the rendering.

Reply:
If any `setTimeout(..., 0)` is done during the render task, or if timers installed before the render task are ready when the render task finishes, they will fire before the `setTimeout`? I believe using setTimeout will run all scheduled/debounced events done in the keypress event for example since setTimeouts callbacks are run in order they are registered (I'm pretty sure).

Reply:
My question is why do we need to invoke `setImmediate` or `setTimeout` in the body of `record`? Here are my suggestions:
* delete `markTextareaSelection` and the state around that
* that leaves a single invocation of `record()`, the one in `markRenderEnd()`
* run all logic synchronously inside `record()`, there's no need to do `setImmediate` or `setTimeout`.

Reply:
If it's synchronously inside `record()` it will mark the end of the input latency measure as before the layout/paint occurs which could significantly lower the expected numbers